File: apps/fountain_cli/test/fountain_cli/credentials_test.exs
defmodule FountainCli.CredentialsTest do
  use ExUnit.Case, async: false

  alias FountainCli.Credentials

  # Use an application-env override so we never touch the real
  # ~/.fountain/credentials during tests.
  setup do
    tmp = Path.join(System.tmp_dir!(), "fountain_test_#{:rand.uniform(999_999)}")
    path = Path.join(tmp, "credentials")
    Application.put_env(:fountain_cli, :credentials_path_override, path)

    on_exit(fn ->
      Application.delete_env(:fountain_cli, :credentials_path_override)
      File.rm_rf!(tmp)
    end)

    {:ok, path: path}
  end

  # ── parse_all/1 ─────────────────────────────────────────────

  describe "parse_all/1" do
    test "parses a single profile" do
      toml = """
      [default]
      api_key = "ftn_abc123"
      base_url = "https://fountain.dev"
      """

      assert Credentials.parse_all(toml) == %{
               "default" => %{
                 "api_key" => "ftn_abc123",
                 "base_url" => "https://fountain.dev"
               }
             }
    end

    test "parses multiple profiles" do
      toml = """
      [default]
      api_key = "ftn_def456"
      base_url = "https://fountain.dev"

      [staging]
      api_key = "ftn_abc123"
      base_url = "https://staging.fountain.dev"
      """

      result = Credentials.parse_all(toml)
      assert result["default"]["api_key"] == "ftn_def456"
      assert result["staging"]["api_key"] == "ftn_abc123"
      assert result["staging"]["base_url"] == "https://staging.fountain.dev"
    end

    test "returns empty map for empty content" do
      assert Credentials.parse_all("") == %{}
    end

    test "ignores blank lines and comment lines" do
      toml = """
      # This is a comment
      [default]
      # Another comment
      api_key = "ftn_abc"
      """

      assert Credentials.parse_all(toml) == %{"default" => %{"api_key" => "ftn_abc"}}
    end

    test "handles values without quotes" do
      toml = """
      [default]
      api_key = ftn_noquotes
      """

      assert Credentials.parse_all(toml)["default"]["api_key"] == "ftn_noquotes"
    end
  end

  # ── read_profile/1 ────────────────────────────────────────

  describe "read_profile/1" do
    test "returns empty map when file does not exist" do
      assert Credentials.read_profile("default") == %{}
    end

    test "returns empty map for a missing profile" do
      Credentials.write_profile("default", %{"api_key" => "ftn_x"})
      assert Credentials.read_profile("nonexistent") == %{}
    end

    test "returns profile attrs when file and profile exist" do
      Credentials.write_profile("default", %{"api_key" => "ftn_abc", "base_url" => "https://fountain.dev"})
      assert Credentials.read_profile("default") == %{
               "api_key" => "ftn_abc",
               "base_url" => "https://fountain.dev"
             }
    end
  end

  # ── write_profile/2 ───────────────────────────────────────

  describe "write_profile/2" do
    test "creates credentials file and parent directories" do
      Credentials.write_profile("default", %{"api_key" => "ftn_abc123"})
      assert File.exists?(Credentials.credentials_path())
    end

    test "written profile can be read back" do
      Credentials.write_profile("default", %{"api_key" => "ftn_abc123", "base_url" => "https://fountain.dev"})
      assert Credentials.read_profile("default") == %{
               "api_key" => "ftn_abc123",
               "base_url" => "https://fountain.dev"
             }
    end

    test "writing staging profile does not alter default profile" do
      Credentials.write_profile("default", %{"api_key" => "ftn_default"})
      Credentials.write_profile("staging", %{"api_key" => "ftn_staging", "base_url" => "https://staging.fountain.dev"})

      assert Credentials.read_profile("default") == %{"api_key" => "ftn_default"}
      assert Credentials.read_profile("staging")["api_key"] == "ftn_staging"
      assert Credentials.read_profile("staging")["base_url"] == "https://staging.fountain.dev"
    end

    test "upsert: updating staging does not alter default" do
      Credentials.write_profile("default", %{"api_key" => "ftn_default"})
      Credentials.write_profile("staging", %{"api_key" => "ftn_staging_v1"})
      Credentials.write_profile("staging", %{"api_key" => "ftn_staging_v2"})

      assert Credentials.read_profile("default") == %{"api_key" => "ftn_default"}
      assert Credentials.read_profile("staging") == %{"api_key" => "ftn_staging_v2"}
    end
  end

  # ── delete_profile/1 ──────────────────────────────────────

  describe "delete_profile/1" do
    test "no-op when file does not exist" do
      assert Credentials.delete_profile("default") == :ok
    end

    test "removes the named section without affecting other profiles" do
      Credentials.write_profile("default", %{"api_key" => "ftn_default"})
      Credentials.write_profile("staging", %{"api_key" => "ftn_staging"})
      Credentials.delete_profile("staging")

      assert Credentials.read_profile("default") == %{"api_key" => "ftn_default"}
      assert Credentials.read_profile("staging") == %{}
    end

    test "no-op when deleting a profile that does not exist in the file" do
      Credentials.write_profile("default", %{"api_key" => "ftn_default"})
      assert Credentials.delete_profile("nonexistent") == :ok
      assert Credentials.read_profile("default") == %{"api_key" => "ftn_default"}
    end
  end

  # ── profile_name/1 ────────────────────────────────────────

  describe "profile_name/1" do
    setup do
      # Ensure FOUNTAIN_PROFILE is clean between tests
      System.delete_env("FOUNTAIN_PROFILE")
      on_exit(fn -> System.delete_env("FOUNTAIN_PROFILE") end)
      :ok
    end

    test "returns the :profile opt when provided" do
      assert Credentials.profile_name(profile: "staging") == "staging"
    end

    test "returns FOUNTAIN_PROFILE env var when no opt" do
      System.put_env("FOUNTAIN_PROFILE", "env_profile")
      assert Credentials.profile_name([]) == "env_profile"
    end

    test "defaults to 'default'" do
      assert Credentials.profile_name([]) == "default"
    end

    test "opt takes precedence over FOUNTAIN_PROFILE env var" do
      System.put_env("FOUNTAIN_PROFILE", "env_profile")
      assert Credentials.profile_name(profile: "opt_profile") == "opt_profile"
    end
  end
end

File: cli/internal/credentials/credentials_test.go
package credentials

import (
	"os"
	"path/filepath"
	"testing"
)

func setup(t *testing.T) string {
	t.Helper()
	tmp := t.TempDir()
	path := filepath.Join(tmp, "credentials")
	SetPathOverride(path)
	t.Cleanup(func() { SetPathOverride("") })
	return path
}

// ── ParseAll ────────────────────────────────────────────────────────────

func TestParseAll_SingleProfile(t *testing.T) {
	toml := `[default]
api_key = "ftn_abc123"
base_url = "https://fountain.dev"
`
	got := ParseAll(toml)
	if got["default"]["api_key"] != "ftn_abc123" {
		t.Fatalf("api_key: %q", got["default"]["api_key"])
	}
	if got["default"]["base_url"] != "https://fountain.dev" {
		t.Fatalf("base_url: %q", got["default"]["base_url"])
	}
}

func TestParseAll_MultipleProfiles(t *testing.T) {
	toml := `[default]
api_key = "ftn_def456"
base_url = "https://fountain.dev"

[staging]
api_key = "ftn_abc123"
base_url = "https://staging.fountain.dev"
`
	got := ParseAll(toml)
	if got["default"]["api_key"] != "ftn_def456" {
		t.Fatalf("default.api_key")
	}
	if got["staging"]["api_key"] != "ftn_abc123" {
		t.Fatalf("staging.api_key")
	}
	if got["staging"]["base_url"] != "https://staging.fountain.dev" {
		t.Fatalf("staging.base_url")
	}
}

func TestParseAll_Empty(t *testing.T) {
	if got := ParseAll(""); len(got) != 0 {
		t.Fatalf("expected empty, got %v", got)
	}
}

func TestParseAll_IgnoresBlanksAndComments(t *testing.T) {
	toml := `# This is a comment
[default]
# Another comment
api_key = "ftn_abc"
`
	got := ParseAll(toml)
	if got["default"]["api_key"] != "ftn_abc" {
		t.Fatalf("got %v", got)
	}
}

func TestParseAll_UnquotedValue(t *testing.T) {
	toml := `[default]
api_key = ftn_noquotes
`
	got := ParseAll(toml)
	if got["default"]["api_key"] != "ftn_noquotes" {
		t.Fatalf("got %q", got["default"]["api_key"])
	}
}

// ── ReadProfile ─────────────────────────────────────────────────────────

func TestReadProfile_NoFile(t *testing.T) {
	setup(t)
	got, err := ReadProfile("default")
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 0 {
		t.Fatalf("expected empty, got %v", got)
	}
}

func TestReadProfile_MissingProfile(t *testing.T) {
	setup(t)
	if err := WriteProfile("default", map[string]string{"api_key": "ftn_x"}); err != nil {
		t.Fatal(err)
	}
	got, err := ReadProfile("nonexistent")
	if err != nil {
		t.Fatal(err)
	}
	if len(got) != 0 {
		t.Fatalf("expected empty, got %v", got)
	}
}

func TestReadProfile_RoundTrip(t *testing.T) {
	setup(t)
	in := map[string]string{"api_key": "ftn_abc", "base_url": "https://fountain.dev"}
	if err := WriteProfile("default", in); err != nil {
		t.Fatal(err)
	}
	got, err := ReadProfile("default")
	if err != nil {
		t.Fatal(err)
	}
	if got["api_key"] != "ftn_abc" || got["base_url"] != "https://fountain.dev" {
		t.Fatalf("got %v", got)
	}
}

// ── WriteProfile ────────────────────────────────────────────────────────

func TestWriteProfile_CreatesParentDirs(t *testing.T) {
	path := setup(t)
	if err := WriteProfile("default", map[string]string{"api_key": "ftn_x"}); err != nil {
		t.Fatal(err)
	}
	if _, err := os.Stat(path); err != nil {
		t.Fatalf("expected file at %s: %v", path, err)
	}
}

func TestWriteProfile_StagingDoesNotAlterDefault(t *testing.T) {
	setup(t)
	_ = WriteProfile("default", map[string]string{"api_key": "ftn_default"})
	_ = WriteProfile("staging", map[string]string{"api_key": "ftn_staging", "base_url": "https://staging.fountain.dev"})

	def, _ := ReadProfile("default")
	if def["api_key"] != "ftn_default" || len(def) != 1 {
		t.Fatalf("default mutated: %v", def)
	}
	stg, _ := ReadProfile("staging")
	if stg["api_key"] != "ftn_staging" || stg["base_url"] != "https://staging.fountain.dev" {
		t.Fatalf("staging: %v", stg)
	}
}

func TestWriteProfile_UpsertDoesNotAlterDefault(t *testing.T) {
	setup(t)
	_ = WriteProfile("default", map[string]string{"api_key": "ftn_default"})
	_ = WriteProfile("staging", map[string]string{"api_key": "ftn_staging_v1"})
	_ = WriteProfile("staging", map[string]string{"api_key": "ftn_staging_v2"})

	def, _ := ReadProfile("default")
	if def["api_key"] != "ftn_default" {
		t.Fatalf("default mutated: %v", def)
	}
	stg, _ := ReadProfile("staging")
	if stg["api_key"] != "ftn_staging_v2" {
		t.Fatalf("staging not upserted: %v", stg)
	}
}

// ── DeleteProfile ───────────────────────────────────────────────────────

func TestDeleteProfile_NoFile(t *testing.T) {
	setup(t)
	if err := DeleteProfile("default"); err != nil {
		t.Fatal(err)
	}
}

func TestDeleteProfile_RemovesNamedSection(t *testing.T) {
	setup(t)
	_ = WriteProfile("default", map[string]string{"api_key": "ftn_default"})
	_ = WriteProfile("staging", map[string]string{"api_key": "ftn_staging"})

	if err := DeleteProfile("staging"); err != nil {
		t.Fatal(err)
	}
	def, _ := ReadProfile("default")
	if def["api_key"] != "ftn_default" {
		t.Fatalf("default mutated: %v", def)
	}
	stg, _ := ReadProfile("staging")
	if len(stg) != 0 {
		t.Fatalf("staging not removed: %v", stg)
	}
}

func TestDeleteProfile_NoOpOnUnknown(t *testing.T) {
	setup(t)
	_ = WriteProfile("default", map[string]string{"api_key": "ftn_default"})
	if err := DeleteProfile("nonexistent"); err != nil {
		t.Fatal(err)
	}
	def, _ := ReadProfile("default")
	if def["api_key"] != "ftn_default" {
		t.Fatalf("default mutated: %v", def)
	}
}

// ── ProfileName ─────────────────────────────────────────────────────────

func TestProfileName_FromOpts(t *testing.T) {
	t.Setenv("FOUNTAIN_PROFILE", "")
	if got := ProfileName(Opts{Profile: "staging"}); got != "staging" {
		t.Fatalf("got %q", got)
	}
}

func TestProfileName_FromEnv(t *testing.T) {
	t.Setenv("FOUNTAIN_PROFILE", "env_profile")
	if got := ProfileName(Opts{}); got != "env_profile" {
		t.Fatalf("got %q", got)
	}
}

func TestProfileName_DefaultFallback(t *testing.T) {
	t.Setenv("FOUNTAIN_PROFILE", "")
	if got := ProfileName(Opts{}); got != "default" {
		t.Fatalf("got %q", got)
	}
}

func TestProfileName_OptsBeatsEnv(t *testing.T) {
	t.Setenv("FOUNTAIN_PROFILE", "env_profile")
	if got := ProfileName(Opts{Profile: "opt_profile"}); got != "opt_profile" {
		t.Fatalf("got %q", got)
	}
}
